avgratecounter.go

```diff
@@ -2,39 +2,46 @@ package ratecounter
 
 import (
 	"strconv"
-	"sync/atomic"
 	"time"
 )
 
 // An AvgRateCounter is a thread-safe counter which returns
 // the ratio between the number of calls 'Incr' and the counter value in the last interval
 type AvgRateCounter struct {
-	hits     int64
-	counter  Counter
+	hits     *RateCounter
+	counter  *RateCounter
 	interval time.Duration
 }
 
 // NewRateCounter Constructs a new AvgRateCounter, for the interval provided
 func NewAvgRateCounter(intrvl time.Duration) *AvgRateCounter {
 	return &AvgRateCounter{
+		hits:     NewRateCounter(intrvl),
+		counter:  NewRateCounter(intrvl),
 		interval: intrvl,
 	}
 }
 
+func (r *AvgRateCounter) WithResolution(resolution int) *AvgRateCounter {
+	if resolution < 1 {
+		panic("AvgRateCounter resolution cannot be less than 1")
+	}
+
+	r.hits = r.hits.WithResolution(resolution)
+	r.counter = r.counter.WithResolution(resolution)
+
+	return r
+}
+
 // Incr Adds an event into the AvgRateCounter
 func (a *AvgRateCounter) Incr(val int64) {
-	atomic.AddInt64(&a.hits, 1)
+	a.hits.Incr(1)
 	a.counter.Incr(val)
-
-	time.AfterFunc(a.interval, func() {
-		atomic.AddInt64(&a.hits, -1)
-		a.counter.Incr(-1 * val)
-	})
 }
 
 // Rate Returns the current ratio between the events count and its values during the last interval
 func (a *AvgRateCounter) Rate() float64 {
-	hits, value := atomic.LoadInt64(&a.hits), a.counter.Value()
+	hits, value := a.hits.Rate(), a.counter.Rate()
```

Comment:
It is true, but this effect is only as bad as resolution is low. The change from actual value can only differ by as much as caused by decreasing ticker in one go in RateCounter itself, so it's not worse. To fix this we'd need to introduce locks or some weird counting technique with history, but I think it's not worth it.

As a future improvement I'd recommend exporting "Tick()" method in RateCounter and doing the counting here instead. I don't think it's that big of an issue though as even if hits and counter are out of sync, they just represent averages, over last interval. They don't need to be 100% perfect.